### PR TITLE
Implemented TFLogSinks

### PR DESCRIPTION
**Pull request by avitebskiy**
- added a `no_default_logger` config setting
- changed `TFLogEntry::log_severity()` to `TFLogEntry::Severity()`
- changed `TFLogEntry::ToString()` to `TFLogEntry::Message()`
- added `TFLogEntry::FName()` and `TFLogEntry::Line()`
- added `TFDefaultLogSink`
- `TFAddLogSink` and `TFRemoveLogSink` now work
- Issue #37390

### CODE REVIEW

**Comment by avitebskiy** [comment]
@googlebot I signed it!

**Comment by allenlavoie** [comment]
Rohan, adding you because it looks like you expressed interest in reviewing in the attached issue. But please comment if that's changed.

Rohan is currently out, so I wouldn't expect an immediate response.

**Comment by gbaned** [comment]
@rohan100jain  Any update on this PR? Please. Thanks!

**Comment by gbaned** [comment]
@rohan100jain  Any update on this PR? Please. Thanks!

**Comment by gbaned** [comment]
@rohan100jain Any update on this PR? Please. Thanks!

**Comment by avitebskiy** [comment]
Come on y'all. This has been pending since July. Any updates?

**Comment by avitebskiy** [comment]
Look, I know this is not the most sophisticated logging scheme: this will work just the same for what it already did and allow users to actually redirect TF logging. It is not an asynchronous logging system, so adding multiple sinks will degrade performance. However, this implements the interface that existed in TF for years and lied about it being implemented the whole time. 

If you have any questions about whether the code is valid, I'll certainly answer them, but I don't understand why this is taking so long to approve and so many people want to look at it before merging. At the end of the day, it doesn't change the existing behavior one bit, and it's a starting point for someone to make this a better logging system in the future if you ever decide to dedicate resources to this. This is one of many patches I already have to apply to our TensorFlow build, please make this one less.

File: tensorflow/core/platform/default/logging.cc
Comment (tabbott) [inline tensorflow/core/platform/default/logging.cc:39]: Please keep the existing API the same (TF{Add,Remove}LogSink). It can forward to your proposed implementation.

File: tensorflow/core/platform/default/logging.cc
Comment (tabbott) [inline tensorflow/core/platform/default/logging.cc:35]: Please use tensorflow/core/platform/mutex.h instead of <mutex> directly. Then replace std::mutex with (tensorflow::)mutex and std::lock_guard with (tensorflow::)mutex_lock.

File: tensorflow/core/platform/default/logging.cc
Comment (tabbott) [inline tensorflow/core/platform/default/logging.cc:51]: Please keep the existing API the same (TF{Add,Remove}LogSink). It can forward to your proposed implementation.

**Comment by evu** [comment]
I too am interested in seeing this make it into tensorflow. This has been needed for a while.

**Comment by avitebskiy** [comment]
Alright, I see the CI now! 

Sorry about the broken build: I was merging this from a patch to 1.15.4, must have made a mistake with that. Will fix today.

**Comment by avitebskiy** [comment]
OK, fixed. Everything should build cleanly now.

File: tensorflow/core/platform/default/logging.cc
Comment (kkimdev) [inline tensorflow/core/platform/default/logging.cc:66]: Adding and removing API takes a raw pointer, e.g., `TFAddLogSink(TFLogSink* sink)` which implies it doesn't take the ownership.  So let's use `std::vector<TFLogSink*>`.

File: tensorflow/core/platform/default/logging.cc
Comment (kkimdev) [inline tensorflow/core/platform/default/logging.cc:59]: This is like `int integer;`.  Let's use more descriptive name e.g., `log_entry_buffer_;`

File: tensorflow/core/platform/default/logging.cc
Comment (kkimdev) [inline tensorflow/core/platform/default/logging.cc:60]: `kMaxLogEntryBufferSize` https://google.github.io/styleguide/cppguide.html#Constant_Names

File: tensorflow/core/platform/default/logging.cc
Comment (kkimdev) [inline tensorflow/core/platform/default/logging.cc:389]: Let's simplify the code path by registering `TFDefaultLogSink` in `TFLogSinks`'s constructor conditionally.  Then we don't need this macro here.

File: tensorflow/core/platform/default/logging.cc
Comment (kkimdev) [inline tensorflow/core/platform/default/logging.cc:366]: If user passes a nullptr, it's likely a bug and it'll be good to be discovered early.  So let's remove the check and let it crash for nullptr.  Same for `TFLogSinks::Remove`.

File: tensorflow/core/platform/default/logging.cc
Comment (kkimdev) [inline tensorflow/core/platform/default/logging.cc:388]: Use `tensorflow::tf_shared_lock` for read access and `tensorflow::mutex_lock` for write access.

File: tensorflow/core/platform/default/logging.h
Comment (kkimdev) [inline tensorflow/core/platform/default/logging.h:494]: nit: let's use -1 as it's more common as an invalid state.

**Review by kkimdev — CHANGES_REQUESTED** [review]
We sincerely apologize for the delay.  I left some comments but the overall direction looks fine, and thank you for the contribution!

File: tensorflow/core/platform/default/logging.h
Comment (avitebskiy) [reply tensorflow/core/platform/default/logging.h:486]: OK, I'll change that. Just looks weird, that's all.

File: tensorflow/tensorflow.bzl
Comment (avitebskiy) [reply tensorflow/tensorflow.bzl:347]: Hmm, that may have been a mistake, I'll take a look.

File: tensorflow/core/platform/default/logging.cc
Comment (avitebskiy) [reply tensorflow/core/platform/default/logging.cc:43]: Will do

File: tensorflow/core/platform/default/logging.cc
Comment (avitebskiy) [reply tensorflow/core/platform/default/logging.cc:362]: Hmm, good to know. Looks like a C++11 change I didn't know about. Will fix.

File: tensorflow/core/platform/default/logging.cc
Comment (avitebskiy) [reply tensorflow/core/platform/default/logging.cc:65]: Will do

File: tensorflow/core/platform/default/logging.cc
Comment (avitebskiy) [reply tensorflow/core/platform/default/logging.cc:60]: Will do

File: tensorflow/core/platform/default/logging.cc
Comment (avitebskiy) [reply tensorflow/core/platform/default/logging.cc:366]: Will, do. That's really not how we do things, but then again we use exceptions, while you don't.

File: tensorflow/core/platform/default/logging.cc
Comment (avitebskiy) [reply tensorflow/core/platform/default/logging.cc:388]: Will do.

File: tensorflow/core/platform/default/logging.h
Comment (avitebskiy) [reply tensorflow/core/platform/default/logging.h:494]: Will do

File: tensorflow/core/platform/default/logging.cc
Comment (avitebskiy) [reply tensorflow/core/platform/default/logging.cc:389]: The default logger is not meant to be part of the queue. It's meant to be a fallback if other log sink is present. I can't just add it to the the queue by default as it has quite severe repercussions. These are the 3 other "solutions" (in the order from simplest to most complicated) I can think of if we did as you suggest: 

* the default logger will always be present with no way for a user to remove it
* we'd have to or skip the first logger in the queue if there's more than one (with the first entry always being the default logger).
* we'd have to remove the default logger the first time another log sink is added. This means that we'd either have to check the first entry to see if it's the default logger or introduce a flag that's set the first time a sink is added. This would also required additional logic to re-add the default logger when the last log sink is removed

After considering the above three solutions, I think I chose the simplest (keep the default logger separate from the queue). In complexity is probably about the same as #2 in the list, but the first one introduces a bug as far as I'm concerned.

File: tensorflow/core/platform/default/logging.h
Comment (avitebskiy) [reply tensorflow/core/platform/default/logging.h:527]: That's not quite true. If you built with `no_default_logger`, but still wanted to use the default `stdout` logger. This implementation would be available for that.

File: tensorflow/core/platform/default/logging.cc
Comment (avitebskiy) [reply tensorflow/core/platform/default/logging.cc:516]: will do

File: tensorflow/core/platform/default/logging.cc
Comment (avitebskiy) [reply tensorflow/core/platform/default/logging.cc:66]: The log sink API must take ownership. It's simply too unsafe otherwise. What is your convention for taking ownership?

File: tensorflow/core/platform/default/logging.cc
Comment (avitebskiy) [reply tensorflow/core/platform/default/logging.cc:59]: I thought the fact that it's in this class is descriptive-enough, but fair enough. It's not a buffer, it's a queue, so I'll specify it as `log_entry_queue_`.

File: tensorflow/core/platform/default/logging.cc
Comment (avitebskiy) [reply tensorflow/core/platform/default/logging.cc:388]: There's no true "read" access in logging: there are potential cases of mutation, so just `tensorflow::mutex_lock`, unless y'all have an `upgrade_lock`.

File: tensorflow/tensorflow.bzl
Comment (avitebskiy) [reply tensorflow/tensorflow.bzl:347]: OK, I looked and that statement is not just about Android, though Android is one of the many options. As far as I can tell, this is the most appropriate place to set the preprocessor symbol.

File: tensorflow/core/platform/default/logging.cc
Comment (avitebskiy) [reply tensorflow/core/platform/default/logging.cc:88]: To elaborate on my decision to use a raw pointer here: the reason I chose to do it this way is because I use the pointer value as the logging sink ID. I wanted the user to be able to retain the pointer value for future use, while the logger would take ownership.

These were my considerations for each case:
* If I used `std::unique_ptr`, the user would have to jump through some hoops to retain the pointer value. 
* We could use a `std::shared_ptr`, which is probably the safest, but shared ownership is not ideal in this case IMHO, though my reasons are mostly academic
* I could actually make your suggested change, because it really doesn't matter whether the pointers are freed or not: their lifetime will be the same as application lifetime. The big problem here is if we give user the ownership, they can really bring the application down in unexpected ways by freeing that pointer.

I'm going to wait on your answer before doing anything on this one, I just wanted you to know the "why". I feel that I made the "least bad" choice here, but I know that your coding standards are drastically different than mine, so my choice may not be the right one for the situation.

**Comment by avitebskiy** [comment]
@kkimdev, I addressed all of your PR comments. The ones I fixed, I marked with :+1: , the ones that I don't think should be changed, I marked with :-1: , and there's one on which I need more feedback after you read my reply, it's marked with :confused:

File: tensorflow/tensorflow.bzl
Comment (avitebskiy) [reply tensorflow/tensorflow.bzl:347]: P.S. The reason I didn't document it there is because the option is already documented, while this statement just calls stuff, it doesn't document anything else. I thought a comment there would be out of place.

**Comment by avitebskiy** [comment]
I'm not clear on why all the tests failed in that build. I'm not familiar with looking through Bazel build logs: the only thing I saw was a timeout, I didn't seen any build failures. I don't know if these verification builds isolate one PR or multiple ones, I also don't know whether test build failures just show up as test failures. Please advise.

**Comment by avitebskiy** [comment]
@kkimdev, would you kick the builds again? It should be good to go now I think. Sorry, I simply don't have the setup to test everything locally.

**Comment by avitebskiy** [comment]
@kkimdev, I fixed everything I saw in the test builds so far. I've never actually built the Android target, so hopefully it's all there now. The only resource I have to get feedback on those changes are your CI builds. I fixed everything I saw in the CI, but I'm working blindly there. Linux builds should hopefully be fine now after I corrected for that Buildifier check.

**Comment by avitebskiy** [comment]
@kkimdev, I don't know if it's expensive to run those builds, but would you kick them again if possible? I think they should all pass or be very close with the latest changes...

File: tensorflow/core/platform/default/logging.cc
Comment (kkimdev) [reply tensorflow/core/platform/default/logging.cc:397]: Could you elaborate what you mean by "severe repercussions"?  Also, I'm fine with introducing `TFGetLogSinks()` function that returns all the added loggers if removing the default logger at runtime is important.

Reasons I prefer treating default logger like any other loggers:
- Macro config is essentially a different build so we won't have a test coverage, better to reduce them.
- Enabling/disabling default logger based on other logger status can be confusing, especially when default logger can be added just like any other loggers, `TFAddLogSink(new TFDefaultLogSink());`.

Comment (avitebskiy) [reply tensorflow/core/platform/default/logging.cc:397]: What I meant by "severe repercussions"  is increase in complexity and/or loss of functionality versus having the "default logger" be separate. However, I think you suggested a way to have it both ways if you will.

Let me elaborate on the reason why the "default logger" functionality is needed at all. The problem we ran into was that the dynamic loading and initialization of CUDA components produced log messages that we couldn't intercept because that code ran before we could actually register a log sink. This means that you get stdout spam whether you want it or not.

The solution was the "default logger". The idea is that when the default logger is present, TensorFlow works as expected. This means that no logging changes are needed anywhere else: everything works as expected. However, if you need to take full control of logging, you can set the build flag to disable the default logger. This way you can install a log sink after some of the logging has already taken place. You won't lose log messages because they're in the queue and will be displayed with the log sink you've installed.

The issue with not having a default logger and requiring the user to install a logger before they see any log messages is that it would break existing expected behavior.

Just for completeness, I thought that maybe having the `no_default_logger` flag be an environment variable could solve our issue, but it won't. It's the same issue as we already have with logging happening before the user has a chance to install a log sink. As far as I know, there's no way to delay this initialization, since it seems to be triggered by DLL load.

We may have a compromise solution though, thanks to your suggestion:

1. Add the `TFGetLogSinks()` function as you suggested.
2. Still have the `no_default_logger` flag, but now it only controls one line of code: adding the default logger in the constructor, so there's no need for big chunks of code to be enabled/disabled with the preprocessor.

Thoughts?

File: tensorflow/core/platform/default/logging.cc
Comment (avitebskiy) [reply tensorflow/core/platform/default/logging.cc:88]: OK, that's what we use as well. Your `TFGetLogSinks()` suggestion alleviates the issue with not having a way to remove an existing log sink. 

The reason I'm so adamant about the user not having the sole ownership of a log sink is because this allows a situation where a user will free a log sink while there's still logging to be done. Usually this happens during a program shutdown and can cause behavior that is difficult to debug. We use `shared_ptr` for our logging, but `unique_ptr` is perfectly acceptable as well.

Your comment about just having a static log sink is not always feasible: there are use cases where a user would want to dynamically add and remove log sinks during execution. In that case, I'd rather the user control the log sink lifetime by actively removing the log sink when it's no longer needed than to have a situation where log sink object has been removed, but the log sink is still registered.

Having said that, it's your library, so I don't have a problem with keeping the `TFLogSink*` signature, but not taking ownership. My preference would be to change the signature to `std::unique_ptr<TFLogSink>` or `std::shared_ptr<TFLogSink>`. Please let me know which way you'd like me to go.

File: tensorflow/core/platform/default/logging.cc
Comment (avitebskiy) [reply tensorflow/core/platform/default/logging.cc:69]: Wops, fixed.

File: tensorflow/core/platform/default/logging.cc
Comment (avitebskiy) [reply tensorflow/core/platform/default/logging.cc:397]: Another thing to think about in the future is to make the "default logger be a "console logger" (or "stdout logger" or something) and move the Android stuff into a separate log sink, which could even be added in a different place and/or context. That whole thing is the most egregious preprocessor include/exclude usage in `logging.cc` and it's just plain ugly. 

That's out of scope for this I think. I modified default/Android logger code as little as I could because I don't have an Android environment to develop against, that's why the build broke a couple of times as I discovered some "copypasta" mistakes I made. It's just a suggestion for future enhancements, a good project for a junior engineer to cut his/her teeth on.

**Comment by avitebskiy** [comment]
I'm seeing build failures on MacOS and Windows, but they don't look like they have anything to do with this PR. Should I just ignore these? Should I merge `master` again?

**Comment by kkimdev** [comment]
> I'm seeing build failures on MacOS and Windows, but they don't look like they have anything to do with this PR. Should I just ignore these? Should I merge `master` again?

Yeah please ignore those failures now.  I'm quite busy today but will try to take a look and comment later today.  Thanks for bearing with us!

**Comment by avitebskiy** [comment]
> > I'm seeing build failures on MacOS and Windows, but they don't look like they have anything to do with this PR. Should I just ignore these? Should I merge `master` again?
> 
> Yeah please ignore those failures now. I'm quite busy today but will try to take a look and comment later today. Thanks for bearing with us!

No worries. I think we're about done. Once you look at my comments and make your decisions I think we can wrap it up pretty soon.

File: tensorflow/core/platform/default/logging.cc
Comment (kkimdev) [reply tensorflow/core/platform/default/logging.cc:88]: Let's keep `TFLogSink*` signature, but please feel free to add `TFAddLogSink(std::unique_ptr<TFLogSink>)` and/or `TFAddLogSink(std::shared_ptr<TFLogSink>)` in addition.  You can keep seperate `std::vector<std::unique_ptr<TFLogSink>>` and/or `std::vector<std::shared_ptr<TFLogSink>>` for `TFRemoveLogSink`.

In this way, we can support any lifetime signatures.

File: tensorflow/core/platform/default/logging.cc
Comment (kkimdev) [reply tensorflow/core/platform/default/logging.cc:397]: - 1.2.3. sounds good.
- Could you add google3/third_party/tensorflow/core/platform/logging_test.cc that tests 1. Adding and removing mock log sink 2. logger queue ?

File: tensorflow/core/platform/default/logging.cc
Comment (avitebskiy) [reply tensorflow/core/platform/default/logging.cc:88]: Sounds good. I don't think having it both ways is a good idea. Managing different ownership models in the same registry is going to lead to some ugly and confusing code. I'm just going to go with your model with the user retaining ownership of the sinks. I did some thinking and I think at the end of the day, there are issues with either model of ownership, so I'll just stick to one and document it. Making things too clever introduces more problems than it solves.

I'll make the final changes some time today.

File: tensorflow/core/platform/default/logging.cc
Comment (avitebskiy) [reply tensorflow/core/platform/default/logging.cc:397]: Sure, the test might be lagging behind the other code changes. I'm a novice at best at using Bazel, so the initial setup will be a learning experience.

Comment (avitebskiy) [reply tensorflow/core/platform/default/logging.cc:397]: Also, I won't be able to write an equivalent test for Android.

Comment (avitebskiy) [reply tensorflow/core/platform/default/logging.cc:397]: Will do

**Comment by kkimdev** [comment]
Please leave a comment when it's ready to be reviewed.  Thanks!

**Comment by avitebskiy** [comment]
I think I've addressed everything now. Ready to be reviewed

**Comment by avitebskiy** [comment]
@kkimdev 
OK, this is hopefully my last checkin for this branch. I've merged `master` and cleaned up the test code a bit. Everything should be there now (unless I missed something).

**Comment by avitebskiy** [comment]
One question, I'll be backporting this into the 1.15 branch, and probably the 2.3 branch. Do you want PRs on those (after this one's wrapped up)?

File: tensorflow/core/platform/logging_test.cc
Comment (kkimdev) [inline tensorflow/core/platform/logging_test.cc:180]: nit: use class and add a member function that returns ss.str()

File: tensorflow/core/platform/logging_test.cc
Comment (kkimdev) [inline tensorflow/core/platform/logging_test.cc:188]: Let's divide this to 2 tests

1. Default sink, guarded with `#if defined(__linux__) && !defined(PLATFORM_POSIX_ANDROID)` )
2. TestSink, all platforms.

Comment (avitebskiy) [reply tensorflow/core/platform/logging_test.cc:188]: I won't be able to test the default sink. What I can do is add another test that simply adds a sink and tests that things get redirected to it, but I won't be able to intercept `stdout`. I guess that's better than nothing.

Comment (avitebskiy) [reply tensorflow/core/platform/logging_test.cc:188]: @kkimdev, it's done.

File: tensorflow/core/platform/default/logging.cc
Comment (avitebskiy) [reply tensorflow/core/platform/default/logging.cc:371]: I disagree, `TFLogSinks::Add` does a whole lot more than just pushing back a log sink. Since this is the constructor, we know that it's thread safe and there are no sinks, or messages in the queue.

File: tensorflow/core/platform/logging_test.cc
Comment (avitebskiy) [reply tensorflow/core/platform/logging_test.cc]: In a test, really? That seems like an overkill. :shrug: I guess I'll pretty it up.

**Comment by avitebskiy** [comment]
@kkimdev, sorry, I didn't notice your other comment, so I made two checkins instead of one. I think you'll need to re-kick the CI.

**Comment by avitebskiy** [comment]
@kkimdev , lall your comments have been addressed. It should all be there now...

**Review by kkimdev — APPROVED** [review]

**Review by gbaned — APPROVED** [review]

**Comment by kkimdev** [comment]
@gbaned I don't see the corresponding CL internally, is this because it didn't pass all the Kokoro tests?

**Comment by kkimdev** [comment]
Update: getting internal reviews now, we should be able to land after getting one more lgtm internally! :)

**Comment by kkimdev** [comment]
@avitebskiy I had to update quite a bit to be consistent with internal logger behavior and just removed linux tests that didn't work internally for some reason, but the functionality should be merged and available now.  In case if you have a follow-up changes, please keep it coming.  Thanks!

**Pull request merged**
